_posts/2018-02-06-unrealiable-network.md

```diff
@@ -0,0 +1,116 @@
+
+# Avoid Making Network Requests as Much as You Can
+
+The best way to avoid getting failing network requests is of course to abandom making them completely.
+
+![Failing network requests meme](/img/unrealiable-network/network_requests_meme.jpg)
+
+Thought this is very efficient, it is not very practical for most of us. Majority of web and mobile apps depend on getting data from servers, and thus, making network requests is mandatory. Still, if we cannot abandom them completely, we can at least reduce the amount of them or optimise them to achieve a good compromise.
+
+# Optimising Network Requests
+
+How can we optimise network requests? First, we should analyse what kind of requests our application is making. The developer tools of [Chrome](https://developer.chrome.com/devtools) and [Firefox](https://developer.mozilla.org/fi/docs/Tools) are very practical for this analysis. On the network tab, you can see all the requests your application is making.
+
+![Chrome Network Tab](/img/unrealiable-network/chrome_network_tab.png)
+
+The way to optimise network requests depends on your application. The following questions can help you to find the requests that could be improved:
+
+**Are you making two separate network requests for saving data and then getting the same data again from the server to update the view?** 
+
+The principle of [responsibility segregation](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation) is a good thing in API design. However, using separate requests for saving and updating the view can be slow and error-prone if either one is successful and the other is not. In this case, on would consider changing the implementation of the data saving API to return the updated data. This way, you can save the data and update the view with one request.
+
+Another possible improvement would be to use optimistic update, in which we assume that most of our network requests are going to be successful, and thus we update the app to the state in which the request has supposedly completed successfully, without really waiting the actual request to complete. Only in the case of failure, we update view back to match the server's answer. This will make the application fell more responsible to the end user, especially in cases when data is edited visually. Possible downside can be that the user is not sure whether the data was actually saved or not, so perhaps inform the user when the data has been truly saved on the server.
```

Comment:
> Another possible improvement would be to use optimistic update, in which we assume that most of our network requests are going to be successful, and thus we update the app to the state in which the request has supposedly completed successfully, without really waiting the actual request to complete.

Another possible improvement would be to use optimistic update: we assume that most of our network requests are going to be successful. After sending the network request we update the app to the state in which the it has supposedly completed successfully, without really waiting the actual request to complete.

> Only in the case of failure, we update view back to match the server's answer.

Only in the case of failure, we reupdate the view to match the server's response.

> fell more responsible

feel more responsive

> Possible downside can be

A possible downside is

> so perhaps inform the user when the data has been truly saved on the server

so perhaps the user should be informed when the data save operation has been verified to have completed.